Cache Store

morses-code opened the pull request:
Creating a cache store with no-op implementation

morses-code committed:
Starting of caching

- Added new route 
- Updated api_test to include new route
- Created Cache interface
- Added cache to CreateSessionsHandlerFunc
- Created sessions now added to cache
- Updated tests for CreateSessionHandlerFunc to check if cache.Set had calls
- Created GetSessionHandlerFunc
- Started tests for GetSessionHandlerFunc
- Added no-op for Cache

morses-code committed:
Changing Get from email to ID

daiLlew committed:
updated handler to take function to get request params

daiLlew commented on api/handlers_test.go:
```diff
@@ -136,6 +139,44 @@ func TestCreateSessionHandlerFunc(t *testing.T) {
 }
 
+func TestGetSessionHandlerFunc(t *testing.T) {
+	Convey("Given a request to retrieve a stored session", t, func() {
```
Think of these statements like BDD statements - they describe the behaviour but don't need to be too technical for example _Given a valid request_

morses-code committed:
GetByID handler finished

- Renamed handler to GetByIDSessionHandlerFunc
- Renamed Cache func to GetByID
- Added INFO logging when Session is stored in Cache
- Created helper func for Marshalling errors
- Added ERROR logging when failed to retrieve Session by ID
- Updated tests with new namings
- Added test to cover calls made when ID isn't valid
- Added helper func for getting request variables

daiLlew commented on api/handlers_test.go:
```diff
@@ -136,6 +139,44 @@ func TestCreateSessionHandlerFunc(t *testing.T) {
+		req := httptest.NewRequest("GET", "/session/123", nil)
+		resp := httptest.NewRecorder()
+
+		Convey("When the request is handled by the router", func() {
```
We arent testing the router here we are testing the handler - so something like _When the request is received_

daiLlew commented on api/handlers_test.go:
```diff
@@ -136,6 +139,44 @@ func TestCreateSessionHandlerFunc(t *testing.T) {
+		Convey("When the request is handled by the router", func() {
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then the correct session details", func() {
```
_Then the expected session is returned_

daiLlew commented on api/handlers.go:
```diff
@@ -52,6 +70,29 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
+		ctx := r.Context()
+		ID := getVarsFunc(r)["ID"]
+
+		log.Event(ctx, fmt.Sprintf("get session by ID: %s", ID), log.INFO)
```
Probably a good idea to remove this as we should n**ever** log out the session ID in production. If someone were able obtain one they would be able access things they shouldn't. Best to delete it now before we forget.

daiLlew commented on api/handlers.go:
```diff
@@ -52,6 +70,29 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
+
+		log.Event(ctx, fmt.Sprintf("get session by ID: %s", ID), log.INFO)
+
+		result, err := cache.Get(ID)
```
You need to add unit tests to cover cases where:
- No session is found.
- Get session returns an error.

morses-code replied:
got one for no session found, not sure about how to cover one for when a get session returns an error.

daiLlew commented on api/handlers.go:
```diff
@@ -52,6 +70,29 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
+		log.Event(ctx, fmt.Sprintf("get session by ID: %s", ID), log.INFO)
+
+		result, err := cache.Get(ID)
+		if err != nil {
```
This needs to return the appropriate status code. If the session is not found 404, otherwise I think a 500 as something went wrong server side.

morses-code replied:
already added the 404 when session isn't found. not sure how to read into the errors to see what type they could be. Is there some sort of switch function I need on it?

daiLlew replied:
In Go its sometimes common to have a public error constant for specific cases i.e. `io.EOF` this enables you to check for specific error cases and allows you to handle them how you like.

At the moment I don't think this is necessary here - if the session is not found we will return `nil, nil`. Its the caller's responsibility to check if the session returned is nil - this should be stated in the Godoc for the `Set()` method making it clear what the expectation is.

It is possible this could change - we might encounter a case where a not found error would be a better design. If that happens we can reassess but for now return nil.

morses-code replied:
> At the moment I don't think this is necessary here - if the session is not found we will return nil, nil. Its the caller's responsibility to check if the session returned is nil - this should be stated in the Godoc for the **Set()** method making it clear what the expectation is.

I'm assuming that `Set()` should be the `GetByID()` function instead as thats the one returning the session and the error?

daiLlew replied:
Yes sorry my mistake. However I think `Set()` should also return an error - It will be using Redis if the client cannot connect or an incorrect status is returned or it dies midway through the request then we would need to propagate that back through the app and treat the request to the sessions API and unsuccessful.

morses-code replied:
Added error return on `Set()` this morning, was thinking exactly as you said.

daiLlew commented on api/handlers.go:
```diff
@@ -52,6 +70,29 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
+		}
+
+		resultJSON, err := result.MarshalJSON()
+		if err != nil {
```
Same here.

daiLlew commented on api/handlers.go:
```diff
@@ -15,8 +18,17 @@ type Sessions interface {
 
+// Cache interface for storing sessions
+type Cache interface {
+	Set(s *session.Session)
```
It might be worthwhile adding an error return type to this method - I think it's a safe bet that the real implementation could run into issues.

daiLlew commented on api/handlers.go:
```diff
@@ -42,8 +54,14 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
+			Email: sess.Email,
+			Start: sess.Start,
+		}
+		cache.Set(s)
```
Add an error return to this method and then add the error checking/status codes etc. You will also need to update the unit tests to account for this scenario.

daiLlew commented on cache/cache.go:
```diff
@@ -0,0 +1,56 @@
+
+var mutex = &sync.Mutex{}
+
+type Cache struct {
```
Godoc on public types/methods

morses-code committed:
Added GoDoc

morses-code committed:
Updated BDD statements, updated http status code

morses-code committed:
Updated handler, cache and tests to address PR comments

- Added error return on Set function
- Added catch for error on Set function
- Added new test for error return on Set function
- Updated mock to include error return on Set function
- Updated no-op with error return on Set function
- Updated cache to actually return an error when it should of

daiLlew commented on api/handlers.go:
```diff
@@ -42,16 +54,55 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
-		sessJson, err := sess.MarshalJSON()
+		if err := cache.Set(s); err != nil {
+			log.Event(ctx, "unable to add session to cache", log.Error(err), log.ERROR)
+			http.Error(w, "unable to add session to cache", http.StatusInternalServerError)
```
you need to `return` here as this will continue through the method.

daiLlew commented on api/handlers.go:
```diff
@@ -42,16 +54,55 @@ func CreateSessionHandlerFunc(sessions Sessions) http.HandlerFunc {
 	}
 }
+
+func failedToMarshalError(ctx context.Context, err error, w http.ResponseWriter) {
+	log.Event(ctx, "failed to marshal session", log.Error(err), log.ERROR)
+	http.Error(w, "Failed to marshal session", http.StatusInternalServerError)
+}
```
I think you could take this a step further to make it more resuable

```
func writeErrorResponse(ctx context.Context, w http.ResponseWriter, msg string, err error, status int) {
	log.Event(ctx, msg, log.Error(err), log.ERROR)
	http.Error(w, msg, status)
}
```
You can this use this for any case where you need to write an unsuccessful response.

daiLlew commented on api/handlers_test.go:
```diff
@@ -134,6 +139,100 @@ func TestCreateSessionHandlerFunc(t *testing.T) {
+				So(mockCache.GetByIDCalls(), ShouldHaveLength, 1)
+			})
+		})
+	})
```
Add some test cases for 
```
result, err := cache.GetByID(ID)
```
- Where result is nil
- error is not nil.

morses-code committed:
Updated handler and tests

- Added return from error when unable to add to cache
- Updated error response when GetByID returns an error
- Added a check on result, if empty the session couldn't be found.
- Updated tests with correct http responses
- Added new test to check nil return session

morses-code committed:
Using a helper for error logging, updated Cache

- Added a new helper for error logging
- Updated Set() func in Cache to add error return

morses-code committed:
Comment update

daiLlew commented on api/handlers.go:
```diff
@@ -15,43 +18,95 @@ type Sessions interface {
 		}
 
 		log.Event(ctx, "session created", log.Data{"session": sess}, log.INFO)
+		s := &session.Session{
```
Don't log out the session - the ID will be in the logs which will be a security vulnerability.

daiLlew commented on api/handlers.go:
```diff
@@ -15,43 +18,95 @@ type Sessions interface {
+			writeErrorResponse(ctx, w, "unable to add session to cache", err, http.StatusInternalServerError)
+			return
+		}
+		log.Event(ctx, "session added to cache", log.INFO)
```
I don't think this ☝️ is needed - it will just add noise to the logs.

daiLlew commented on api/handlers.go:
```diff
@@ -15,43 +18,95 @@ type Sessions interface {
+	if err != nil {
+		log.Event(ctx, msg, log.Error(err), log.ERROR)
 	}
+	log.Event(ctx, msg, log.ERROR)
```
I think this needs to be in an `else` otherwise you will log the message twice.

morses-code committed:
Removed some logging and corrected if statement

jessjenkins commented on api/handlers.go:
```diff
@@ -15,43 +18,95 @@ type Sessions interface {
+	}
+}
+
+func GetByIDSessionHandlerFunc(cache Cache, getVarsFunc GetVarsFunc) http.HandlerFunc {
```
Should have GoDoc if this is an exported function. 
```suggestion
// GetByIDSessionHandlerFunc does this that or the other
func GetByIDSessionHandlerFunc(cache Cache, getVarsFunc GetVarsFunc) http.HandlerFunc {
```

jessjenkins commented on api/api_test.go:
```diff
@@ -3,10 +3,9 @@ package api
 import (
 	"context"
 	"github.com/gorilla/mux"
+	. "github.com/smartystreets/goconvey/convey"
```
Small suggestion but it's good to import the standard go libraries before external dependencies. 
The imports in `api/handlers.go` is a good example.

jessjenkins commented on api/handlers.go:
```diff
@@ -15,43 +18,95 @@ type Sessions interface {
+		log.Event(ctx, msg, log.ERROR)
+	}
+	http.Error(w, msg, status)
+
```
Tiny: no need for extra newline here.

jessjenkins commented on api/nop.go:
```diff
@@ -5,15 +5,42 @@ import (
+	return nil
+}
+
+func (n *NOPCache) GetByID(ID string) (*session.Session, error) {
```
Needs GoDoc

jessjenkins commented on api/nop.go:
```diff
@@ -5,15 +5,42 @@ import (
 	return sess, nil
 }
+
+func (n *NOPCache) Set(s *session.Session) error {
```
Needs GoDoc

morses-code committed:
Updated GoDoc

- Reordered imports in api_test
- Added GoDoc to GetByIDSessionHandlerFunc()
- Added GoDoc to no-op functions
